File: coffee-filter-ui/app/lib/geocoding.ts
export interface Coordinates {
  latitude: number;
  longitude: number;
}

/**
 * Geocodes an address to latitude and longitude coordinates.
 * Uses OpenStreetMap Nominatim API (free, no API key required).
 *
 * Note: Nominatim has rate limits (1 request per second recommended).
 * For production use, consider caching results or using a paid service.
 *
 * @param address - The address string to geocode (e.g., "4141 Troost Ave, Kansas City, MO 64110")
 * @returns Promise with coordinates object, or null if geocoding fails
 */
export async function geocodeAddress(
  address: string
): Promise<Coordinates | null> {
  try {
    const encodedAddress = encodeURIComponent(address);
    const url = `https://nominatim.openstreetmap.org/search?q=${encodedAddress}&format=json&limit=1`;

    const response = await fetch(url, {
      headers: {
        "User-Agent": "CoffeeFilter/1.0", // Required by Nominatim ToS
      },
    });

    if (!response.ok) {
      throw new Error(`Geocoding failed: ${response.statusText}`);
    }

    const data = await response.json();

    if (!data || data.length === 0) {
      return null;
    }

    const result = data[0];
    return {
      latitude: parseFloat(result.lat),
      longitude: parseFloat(result.lon),
    };
  } catch (error) {
    console.error("Error geocoding address:", error);
    return null;
  }
}

File: backend/app/core/geocoding.py
"""
Geocoding utility using OpenStreetMap Nominatim API.
"""
import httpx
from urllib.parse import quote
from typing import Optional, Tuple


async def geocode_address(address: str) -> Optional[Tuple[float, float]]:
    """
    Geocode an address to latitude and longitude coordinates.
    Uses OpenStreetMap Nominatim API (free, no API key required).
    
    Args:
        address: The address string to geocode
        
    Returns:
        Tuple of (latitude, longitude) if successful, None otherwise
    """
    try:
        encoded_address = quote(address)
        url = f"https://nominatim.openstreetmap.org/search?q={encoded_address}&format=json&limit=1"
        
        async with httpx.AsyncClient() as client:
            response = await client.get(
                url,
                headers={
                    "User-Agent": "CoffeeFilter/1.0",  # Required by Nominatim ToS
                },
                timeout=10.0
            )
            
            if response.status_code != 200:
                return None
                
            data = response.json()
            
            if not data or len(data) == 0:
                return None
                
            result = data[0]
            latitude = float(result["lat"])
            longitude = float(result["lon"])
            
            return (latitude, longitude)
            
    except Exception as e:
        print(f"Error geocoding address: {e}")
        return None

